compiler/rustc_mir/src/transform/check_consts/validation.rs

```diff
@@ -556,7 +582,25 @@ impl Visitor<'tcx> for Validator<'mir, 'tcx> {
                 );
 
                 if borrowed_place_has_mut_interior {
-                    self.check_op(ops::CellBorrow);
+                    match self.const_kind() {
+                        // In a const fn all borrows are transient or point to the places given via
+                        // references in the arguments. The borrow checker guarantees that.
+                        // NOTE: Once we have heap allocations during CTFE we need to figure out
+                        // how to prevent `const fn` to create long-lived allocations that point
+                        // to (interior) mutable memory.
+                        hir::ConstContext::ConstFn => self.check_op(ops::TransientCellBorrow),
+                        _ => {
+                            // Locals without StorageDead follow the "enclosing scope" rule, meaning
+                            // they are essentially anonymous static items themselves.
```

Comment:
The comment is focusing on the wrong case -- it should explain why locals *with* `StorageDead` are okay to treat as "definitely transient".